### PR TITLE
Enable extended FAB hero transition resizing

### LINKED ISSUE #18782: FloatingActionButton.extended causes RenderFlex errors
Issue #17583 has been closed however it still happens for me and I have found that if you null out the hero parameter as shown below you stop getting the RenderFlex errors. I can't see a way that we can re-open a closed issue?

```
floatingActionButton: FloatingActionButton.extended(
        onPressed: _createNotification,
        icon: Icon(Icons.add),
        label: const Text('New'),
        heroTag: null,
      ),
```

According to the documentation, the default value for hero tag should align with the navigated-to screens hero (so long as there is only one on each screen). It does do this properly however on the nav back you get the renderflex error and very briefly see the yellow and black hatch pattern on the right-side of the floating action button.

## Steps to Reproduce

1. Use the extended form of the floating action button and on click (as shown a above)- navigate to a new screen
2. Go back (pop) and RenderFlex errors will be displayed in the console UNLESS you set the heroTag to null.

## Logs

```
Performing hot reload...
Reloaded 1 of 396 libraries in 1,695ms.
flutter: ══╡ EXCEPTION CAUGHT BY RENDERING LIBRARY ╞═════════════════════════════════════════════════════════
flutter: The following message was thrown during layout:
flutter: A RenderFlex overflowed by 45 pixels on the right.
flutter:
flutter: The overflowing RenderFlex has an orientation of Axis.horizontal.
flutter: The edge of the RenderFlex that is overflowing has been marked in the rendering with a yellow and
flutter: black striped pattern. This is usually caused by the contents being too big for the RenderFlex.
flutter: Consider applying a flex factor (e.g. using an Expanded widget) to force the children of the
flutter: RenderFlex to fit within the available space instead of being sized to their natural size.
flutter: This is considered an error condition because it indicates that there is content that cannot be
flutter: seen. If the content is legitimately bigger than the available space, consider clipping it with a
flutter: ClipRect widget before putting it in the flex, or using a scrollable container rather than a Flex,
flutter: like a ListView.
flutter: The specific RenderFlex in question is:
flutter:   RenderFlex#ba856 relayoutBoundary=up1 OVERFLOWING
flutter:   creator: Row ← IconTheme ← Builder ← Center ← Padding ← Container ← IconTheme ← Builder ← Listener
flutter:   ← _GestureSemantics ← RawGestureDetector ← GestureDetector ← ⋯
flutter:   parentData: offset=Offset(0.0, 16.0) (can use size)
flutter:   constraints: BoxConstraints(0.0<=w<=56.3, 0.0<=h<=56.0)
flutter:   size: Size(56.3, 24.0)
flutter:   direction: horizontal
flutter:   mainAxisAlignment: start
flutter:   mainAxisSize: min
flutter:   crossAxisAlignment: center
flutter:   textDirection: ltr
flutter:   verticalDirection: down
flutter: ◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤
flutter: ════════════════════════════════════════════════════════════════════════════════════════════════════

```

No issues shown in flutter doctor nor flutter analyze.

### PR DESCRIPTION
**Pull request by HansMuller**
Fixes https://github.com/flutter/flutter/issues/18782

All FAB's are heroes by default and extended FABs (new FloatingActionButton.extended) are built from a icon,label Row with MainAxisSize.min. So the widget of an extended FAB depends on the widths of its label and icon.

At the beginning of a hero-transition from an extended FAB with a short label like 'X' to one with a long label like 'I am a long FAB label', we try to size the long extended FAB to match the short one. This fails because MainAxisSize.min: the width of the underlying RenderFlex can't be smaller than the sum of the widths of the label and icon.

Added a _ChildOverflowBox widget whose size matches its child's size unless its constraints force it to be larger or smaller. The child is centered. Extended FABs wrap their Row with the _ChildOverflowBox  The result is clipped because the extended FAB's Row is enclosed by material shape.

### CODE REVIEW

File: packages/flutter/lib/src/material/floating_action_button.dart [hunk 1]
```diff
@@ -273,3 +278,68 @@ class _FloatingActionButtonState extends State<FloatingActionButton> {
...
+
+// This widget's size matches its child's size unless its constraints
+// force it to be larger or smaller. The child is centered.
+//
+// Used to encapsulate extended FABs whose size is fixed, using Row
+// and MainAxisSize.min, to be as wide as their label and icon.
+class _ChildOverflowBox extends SingleChildRenderObjectWidget {
+  const _ChildOverflowBox({
+    Key key,
+    this.alignment = Alignment.center,
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/material/floating_action_button.dart]: Since you never set center would it be better to hard code it?

**Review by jonahwilliams — APPROVED** [review]
LGTM with nit

**Comment by creativecreatorormaybenot** [comment]
@HansMuller This is merged, but I am still having the issue. Is there any additional action I need to take?

**Comment by HansMuller** [comment]
@creativecreatorormaybenot - the bug is fixed in the `master` branch (like tip of tree) of the flutter repo. I don't believe that it's been included in the beta channel yet.

**Pull request merged**